index.js

```diff
@@ -7,555 +7,103 @@
- *
- * @param  {Object} `err` Error object
- * @param  {Object} `helper` helper object to provide more information
- * @param  {Array} `args` Array of arguments passed to the helper.
- * @return {Object} Formatted Error object
- */
-
-function formatError(err, helper, args) {
-  err.helper = helper;
-  define(err, 'args', args);
-  return err;
-}
-
-/**
- * Create a prefix to use when generating an async id.
- *
- * @param  {String} `prefix` prefix string to start with.
- * @param  {String} `counter` string to append.
- * @return {String} new prefix
- */
+  async resolve(str) {
+    const self = this;
+    const promises = Object.keys(this.ids)
+      .map((id) => this.ids[id])
+      .map(async({ id, args, fn, context }) => {
 
-function appendPrefix(prefix, counter) {
-  return prefix + counter + '$';
-}
+        const argz = args.map(function func(arg) {
+          const item = self.ids[arg];
+          return item
+            ? item.fn.call(item.context, ...item.args.map(func))
+            : arg;
+        });
 
-/**
- * Create an async id from the provided prefix and counter.
- *
- * @param  {String} `prefix` prefix string to start with
- * @param  {String} `counter` string to append.
- * @return {String} async id
- */
-
-function createId(prefix, counter) {
-  return appendPrefix(prefix, counter) + '}';
-}
-
-/**
- * Create a regular expression based on the given `prefix`.
- * @param  {String} `prefix`
- * @return {RegExp}
- */
-
-function toRegex(prefix) {
-  var key = appendPrefix(prefix, '(\\d+)');
-  if (cache.hasOwnProperty(key)) {
-    return cache[key];
-  }
-  var regex = new RegExp(createRegexString(key), 'g');
-  cache[key] = regex;
-  return regex;
-}
+        this.ids[id].value = await fn.call(context, ...(await Promise.all(argz)));
 
-/**
- * Create a string to pass into `RegExp` for checking for and finding async ids.
- * @param  {String} `prefix` prefix to use for the first part of the regex
- * @return {String} string to pass into `RegExp`
- */
+        return this.ids[id];
+      });
 
-function createRegexString(prefix) {
-  var key = 'createRegexString:' + prefix;
-  if (cache.hasOwnProperty(key)) {
-    return cache[key];
+    return Promise.all(promises).then((res) =>
+      res.reduce(
+        (acc, { id, value }) => {
+          if (acc === id) return value;
+          return acc.replace(new RegExp(escapeRegex(id), 'g'), value);
+        },
+        str,
+      ),
+    );
   }
-  var str = (prefix + '(\\d+)$}').replace(/\\?([${}])/g, '\\$1');
-  cache[key] = str;
-  return str;
 }
 
-/**
- * Return true if the given value is a helper "group"
- */
-
-function isHelperGroup(helpers) {
-  if (!helpers) return false;
-  if (helpers.isGroup) {
-    return true;
-  }
-  if (typeof helpers === 'function' || isObject(helpers)) {
-    var keys = Object.keys(helpers).filter(function(name) {
-      return ['async', 'sync', 'displayName'].indexOf(name) === -1;
+function promisify(fn) {
+  return function func(...args) {
```

Comment:
we should use `util.promisify`

Reply:
Yea, thought about that, but if someone wants to use it in the browser this way may save a lot.

Reply:
That's the reason i didn't used the `assert` too.